README.md
[ ![Codeship Status for jbdoumenjou/cuisinons](https://codeship.com/projects/cadd92d0-5870-0132-763f-2264a2250d8e/status?branch=develop)](https://codeship.com/projects/50079)
[![Deploy to Heroku](https://www.herokucdn.com/deploy/button.png)](https://cuisinons.herokuapp.com/)
# cuisinons
A simple cook recipe API and a technical experimentation

# Test
run npm test command (use mocha)


changelog.md
# 0.2.0
* add root controller (avoid 404 on root get)

# 0.1.0
* create the project
* Basic crud operations on the recipes resource
* CI: gitlab -> codeship -> heroku

test/api.js
'use strict'

const http = require('http')
const expect = require('chai').expect
const server = require('../src/server.js')
const recipeController = require('../src/recipesController')
const fixture = require('./fixtures')

function getPostOption() {
    return {
        host: server.host,
        port: server.port,
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        }
    }
}

function checkRecipe(expected, actual) {
    expect(actual.title).to.deep.equal(expected.title)
    expect(actual.description).to.deep.equal(expected.description)
    expect(actual.language).to.deep.equal(expected.language)
    expect(actual.ingredients).to.deep.equal(expected.ingredients)
    expect(actual.steps).to.deep.equal(expected.steps)
    expect(actual.createdAt).to.exist
    expect(actual.updatedAt).to.exist
    expect(actual.id).to.exist
}

describe('GET /recipes', function () {
    before(() => server.start())
    after(() => server.stop())

    it('should return status code 200', function (done) {
        const url = server.url + "/recipes"
        http.get(url, (res) => {
            expect(res.statusCode).to.be.equal(200)
            done()
        })
    })

    it('should return empty list where there is no recipe', function (done) {
        const url = server.url + "/recipes"
        http.get(url, (res) => {
            let body = []

            res.on('error', function (err) {
                console.error(err);
            }).on('data', function (chunk) {
                body.push(chunk);
            }).on('end', function () {
                body = Buffer.concat(body).toString();
                body = JSON.parse(body)
                expect(body).to.be.instanceof(Array);
                expect(body).to.be.empty
                done()
            });


        })
    })

    it('should return an array of recipes', function (done) {
        const testRecipe = fixture[0]
        recipeController.create(testRecipe)

        const url = server.url + "/recipes"
        http.get(url, (res) => {
            let body = []

            res.on('error', function (err) {
                console.error(err);
            }).on('data', function (chunk) {
                body.push(chunk);
            }).on('end', function () {
                body = Buffer.concat(body).toString()
                body = JSON.parse(body)
                expect(body).to.be.instanceof(Array)
                checkRecipe(testRecipe, body[0])

                done()
            });
        })
    })

})

describe('GET /recipes/:id', function () {
    before(() => server.start())
    after(() => server.stop())

    it('should return status code 200', function (done) {
        let recipePromise = recipeController.create({title: 'myRecipe'})
        let postOption = getPostOption()

        recipePromise.then((recipe)=> {
            postOption.path = '/recipes/' + recipe.id
            postOption.method = 'GET'
            delete postOption.headers

            const request = http.request(postOption, (res) => {
                expect(res.statusCode).to.be.equal(200)
                done()
            })

            request.end()
        })
    })

    it('should return the right recipe', function (done) {
        let recipePromise = recipeController.create({title: 'myRecipe'})

        recipePromise.then((recipe)=> {
            let postOption = getPostOption()
            postOption.path = '/recipes/' + recipe.id
            postOption.method = 'GET'
            delete postOption.headers

            const request = http.request(postOption, (res) => {
                let body = []

                res.on('error', function (err) {
                    console.error(err);
                }).on('data', function (chunk) {
                    body.push(chunk);
                }).on('end', function () {
                    body = Buffer.concat(body).toString()
                    body = JSON.parse(body)
                    checkRecipe(recipe, body)
                    done()
                });
            })

            request.end()
        })
    })
})

describe('POST /recipes', function () {
    before(() => server.start())
    after(() => server.stop())

    it('should return status code 200', function (done) {
        const recipe = fixture[0]
        const data = JSON.stringify(recipe)
        let postOption = getPostOption()
        postOption.path = '/recipes'
        postOption.headers['Content-Length'] = Buffer.byteLength(data)

        const request = http.request(postOption, (res) => {
            expect(res.statusCode).to.be.equal(200)
            done()
        })

        request.write(data)
        request.end()
    })

    it('should return the created recipe', function (done) {
        const recipe = fixture[0]
        const data = JSON.stringify(recipe)
        let postOption = getPostOption()
        postOption.path = '/recipes'
        postOption.headers['Content-Length'] = Buffer.byteLength(data)

        const request = http.request(postOption, (res) => {
            let body = []

            res.on('error', function (err) {
                console.error(err);
            }).on('data', function (chunk) {
                body.push(chunk);
            }).on('end', function () {
                body = Buffer.concat(body).toString()
                body = JSON.parse(body)
                checkRecipe(recipe, body)
                done()
            });
        })

        request.write(data)
        request.end()
    })

})

describe('DELETE /recipes/:id', function () {
    before(() => server.start())
    after(() => server.stop())

    it('should return status code 200', function (done) {
        let recipePromise = recipeController.create({title: 'myRecipe'})

        recipePromise.then((recipe)=> {
            let postOption = getPostOption()
            postOption.path = '/recipes/' + recipe.id
            postOption.method = 'DELETE'
            delete postOption.headers

            const request = http.request(postOption, (res) => {
                expect(res.statusCode).to.be.equal(200)
                done()
            })

            request.end()
        })
    })

    it('should return the deleted recipe id', function (done) {
        let recipePromise = recipeController.create({title: 'myRecipe'})

        recipePromise.then((recipe)=> {
            let postOption = getPostOption()
            postOption.path = '/recipes/' + recipe.id
            postOption.method = 'DELETE'
            delete postOption.headers

            const request = http.request(postOption, (res) => {
                let body = []

                res.on('error', function (err) {
                    console.error(err);
                }).on('data', function (chunk) {
                    body.push(chunk);
                }).on('end', function () {
                    body = Buffer.concat(body).toString()
                    body = JSON.parse(body)
                    expect(body).to.be.equal(recipe.id)
                    done()
                });
            })

            request.end()
        })
    })

})



test/test.js
'use strict'

const http = require('http')
const expect = require('chai').expect
const server = require('../src/server.js')

describe('GET /recipes', function() {
    before(() => server.start())
    after(() => server.stop())

    it('should return status code 200', function (done) {
        const url = server.url + "/recipes"
        http.get(url, (res) => {
            expect(res.statusCode).to.be.equal(200)
            done()
        })
    })
    
})



test/routes.js
'use strict'
const url = require('url')
const routes = require('../src/routes')

describe('test routing of /recipes/rs8zajrlik9)', function () {
    let request = {
        //url: '/recipes/rs8zajrlik9',
        url: '/recipes',
        method: 'POST'
    }
    routes.getCrudAction(request, {title:'test'})
})

/*describe('url.parse(http://localhost:8080/recipe)', function () {

 const request = {
 url: 'http://localhost:8080/recipe/1',
 path: 'recipes/42',
 method: 'GET'
 }

 it('should work', function(){
 let result = url.parse('http://localhost:8080/recipe/1')
 console.log(result)
 })

 })*/

src/routes.js
'use strict'

const recipesController = require('./recipesController')
const rootController = require('./rootController')

const availableResources = {
    '': rootController,
    'recipes': recipesController
}

// basic path parsing
function getResourceInfo(path) {
    let parts = path.split('/')
    if (parts[0].length === 0) {
        parts.shift()
    }

    let resource = parts[0]
    let id = null

    if (parts[1]) {
        id = parts[1]
    }

    return {resource: resource, id: id}
}

/*
 get /resources              => list
 get /resources/:id          => get a specific resource
 post /resources             => create a resource
 patch/put /resources/:id    => update a resource
 delete /resources/:id       => delete a resource
 */
function route(request, body) {
    const method = request.method.toLowerCase()
    const path = request.url

    const result = new Promise((resolve, reject) => {
        let resourceInfo = getResourceInfo(path)
        let controller = availableResources[resourceInfo.resource]
        if (controller) {
            let id = resourceInfo.id
            let controllerPromise = null

            // get /resources => list resources
            if (method === 'get' && !id) {
                controllerPromise = controller.list()
            }

            // get /resources/:id => get resources with id
            if (method === 'get' && id) {
                controllerPromise = controller.get(id)
            }

            // post /resources/ => create a resource
            if (method === 'post' && !id) {
                controllerPromise = controller.create(body)
            }

            // put /resources/:id => edit a specific resource
            if (method === 'put' && id) {
                controllerPromise = controller.update(id, body)
            }

            // delete /resources/:id => delete a specific resource
            if (method === 'delete' && id) {
                controllerPromise = controller.remove(id)
            }

            if (!controllerPromise) {
                reject()
            }
            resolve(controllerPromise)
        } else {
            reject()
        }
    })

    return result
}

module.exports = {
    route: route
}

src/recipesStore.js
'use strict'
// Dummy implementation of a storage
const uuidGenerator = require('../lib/uuid.js')

// TODO use database instead of a memory object
let recipeById = new Map();

function list() {
    let result = new Promise((resolve, reject) => {
        let recipes = Array.from(recipeById, elt => elt[1])

        resolve(recipes)
    })

    return result
}

// get the first unused uid
function getUuid() {
    let uid = uuidGenerator.create()
    while (recipeById.has(uid)) {
        uid = uuidGenerator.create()
    }

    return uid
}

function create(recipe) {
    let result = new Promise((resolve, reject) => {
        recipe.id = getUuid()
        recipe.createdAt = new Date()
        recipe.updatedAt = new Date()
        recipeById.set(recipe.id, recipe)
        resolve(recipe)
    })

    return result
}

function update(recipe) {
    let result = new Promise((resolve, reject) => {
        let old = recipeById.get(recipe.id)
        recipe.id = old.id
        recipe.createdAt = old.createdAt
        recipe.updatedAt = new Date()

        recipeById.set(recipe.id, recipe)
        resolve(recipe)
    })

    return result
}

function get(recipeId) {
    let result = new Promise((resolve, reject) => {
        if (recipeById.has(recipeId)) {
            resolve(recipeById.get(recipeId))
        } else {
            reject('Does not exist')
        }

    })

    return result
}

function remove(recipeId) {
    let result = new Promise((resolve, reject) => {
        recipeById.delete(recipeId)
        resolve(recipeId)
    })

    return result
}

module.exports = {
    list: list,
    create: create,
    get: get,
    update: update,
    remove: remove
}

src/recipesController.js
'use strict'
// Dummy implementation of a storage
const uuidGenerator = require('../lib/uuid.js')

// TODO use database instead of a memory object
let recipeById = new Map();

function list() {
    let result = new Promise((resolve, reject) => {
        let recipes = Array.from(recipeById, elt => elt[1])

        resolve(recipes)
    })

    return result
}

// get the first unused uid
function getUuid() {
    let uid;
    
    do {
        uid = uuidGenerator.create()
    } while (recipeById.has(uid))

    return uid
}

function create(recipe) {
    let result = new Promise((resolve, reject) => {
        recipe.id = getUuid()
        recipe.createdAt = new Date()
        recipe.updatedAt = new Date()
        recipeById.set(recipe.id, recipe)
        resolve(recipe)
    })

    return result
}

function update(id, recipe) {
    let result = new Promise((resolve, reject) => {
        let old = recipeById.get(id || recipe.id)
        recipe.id = old.id
        recipe.createdAt = old.createdAt
        recipe.updatedAt = new Date()

        recipeById.set(recipe.id, recipe)
        resolve(recipe)
    })

    return result
}

function get(recipeId) {
    let result = new Promise((resolve, reject) => {
        if (recipeById.has(recipeId)) {
            resolve(recipeById.get(recipeId))
        } else {
            reject('Does not exist')
        }

    })

    return result
}

function remove(recipeId) {
    let result = new Promise((resolve, reject) => {
        recipeById.delete(recipeId)
        resolve(recipeId)
    })

    return result
}

module.exports = {
    list: list,
    create: create,
    get: get,
    update: update,
    remove: remove
}

src/server.js
'use strict'

const http = require('http')
const routes = require('./routes.js')
const port = process.env.PORT || 5000

let instance

const server = http.createServer((request, response) => {
    const headers = request.headers;
    const method = request.method;
    const url = request.url;
    let body = [];

    request.on('error', function(err) {
        console.error(err);
    }).on('data', function(chunk) {
        body.push(chunk);
    }).on('end', function() {
        if (body.length > 0) {
            body = Buffer.concat(body).toString();
            body = JSON.parse(body)
        }
        routes.route(request, body)
            .then((data) => ok(response, data))
            .catch((data) => fail(response))    
    });

    
})

function ok(response, data) {
    response.writeHead(200, {'Content-Type': 'application/json'})
    let json = JSON.stringify(data)
    response.end(json)
}

function fail(response) {
    response.statusCode = 404
    response.end(http.STATUS_CODES[404])
}

function start() {
    instance = server.listen(port)
    console.log('start listening on port ' + port)
}

function stop() {
    instance.close()
}

module.exports = {
    url: 'http://localhost:' + port,
    host: 'localhost',
    port: port,
    start: start,
    stop: stop
}